memberlist.go

```diff
@@ -191,50 +191,111 @@ func (m *Memberlist) Join(existing []string) (int, error) {
 		}
 
 	}
-
 	if numSuccess > 0 {
 		errs = nil
 	}
 	return numSuccess, errs
 }
 
+// ipPort holds information about a node we want to try to join.
+type ipPort struct {
+	ip   []byte
+	port uint16
+}
+
+// tcpLookupIP is a helper to initiate a TCP-based DNS lookup for the given host.
+// The built-in Go resolver will do a UDP lookup first, and will only use TCP if
+// the response has the truncate bit set, which isn't common on DNS servers like
+// Consul's. By doing the TCP lookup directly, we get the best chance for the
+// largest list of hosts to join. Since joins are relatively rare events, it's ok
+// to do this rather expensive operation.
+func (m *Memberlist) tcpLookupIP(host string, defaultPort uint16) ([]ipPort, error) {
+	cc, err := dns.ClientConfigFromFile(m.config.DNSConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(cc.Servers) > 0 {
+		server := cc.Servers[0]
+		if !hasPort(server) {
+			server = net.JoinHostPort(server, cc.Port)
+		}
+
+		c := new(dns.Client)
+		c.Net = "tcp"
+		m := new(dns.Msg)
+		m.SetQuestion(host, dns.TypeANY)
```

Comment:
This comment is purely informational as there's nothing better to be done, but it's sad that this was brought up as a problem in '98 and it is still a problem (despite repeated attempts to fix over the years).  Item 4 in the following has additional detail.

http://tools.ietf.org/html/draft-sajitha-dnsext-qtype-addr-00

TL;DR: `ANY` is as good as it gets and works for the `A` and `AAAA` QTYPES, but it won't give the DNS server the opportunity to resolve `CNAME` records and return the correct information in the `ADDITIONAL` section.  Hold this thought for a second.

`T_ANY` also has the server xmit all records to the client, which shouldn't be common, and as you said, this is reasonably rare.